### PR TITLE
Psalm 5.4 upgrade

### PR DESCRIPTION
**Pull request by drupol**
This PR:

- [x] Fix issues introduced by Psalm 5.4
- [x] Has static analysis tests (psalm, phpstan)
- [x] Needs PSalm > 5.4.0 to work (_use dev version in the meantime_)

### CODE REVIEW

**Comment by drupol** [comment]
This PR needs to be merged once the upcoming version of PSalm is out (5.4.1).

File: composer.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
     ],
     "require": {
         "php": ">= 8",
-        "loophp/iterators": "^2.3.1"
+        "loophp/iterators": "^2.3.1",
+        "vimeo/psalm": "dev-master#3560ef771743048302599aea151e206a50f4a744"
```
Comment (AlexandruGG) [inline composer.json:41]: Psalm 5.4 is out now, is this still needed? 

https://github.com/vimeo/psalm/releases/tag/5.4.0

File: src/Contract/Operation/Unfoldable.php [hunk 2]
```diff
@@ -14,11 +14,12 @@ interface Unfoldable
      * @see https://loophp-collection.readthedocs.io/en/stable/pages/api.html#unfold
      *
      * @template T
+     * @template TKey
```
Comment (AlexandruGG) [inline src/Contract/Operation/Unfoldable.php:17]: `TKey` is unused here, did you mean to change the return type on line 22 to `Collection<TKey, iterable<array-key, T>>`?

File: src/Contract/Operation/Unwrapable.php [hunk 3]
```diff
@@ -18,7 +18,7 @@ interface Unwrapable
      *
      * @see https://loophp-collection.readthedocs.io/en/stable/pages/api.html#unwrap
      *
-     * @return Collection<mixed, mixed>
+     * @return Collection<TKey, T>
```
Comment (AlexandruGG) [inline src/Contract/Operation/Unwrapable.php:21]: not too sure about this change -> it's not a given that the elements in the unwrapped collection will be all of the same type 🤔 

```PHP
Collection::fromIterable([['a'], [1], [true]])
    ->unwrap(); // ['a', 1, true]
```

unless static analysis is able to interpret the template `T` as the union `string|int|bool` in this case. I'm not sure but you could test this with a static analysis check

File: tests/static-analysis/unwrap.php [hunk 4]
```diff
@@ -8,27 +8,71 @@
...
+ *
+ * @phpstan-param CollectionInterface<int, int> $collection
  */
-function unwrap_checkList(CollectionInterface $collection): void
+function unwrap_checkList1(CollectionInterface $collection): void
 {
 }
 /**
- * @param CollectionInterface<string, string> $collection
+ * @psalm-param CollectionInterface<int<0, 2>, 2|3|list{1}> $collection
```
Comment (AlexandruGG) [inline tests/static-analysis/unwrap.php:19]: this is not good, and it shows that the new type hints have made `unwrap` too restrictive and incorrect in this case

```PHP
Collection::fromIterable([[1], 2, 3])->unwrap() 
// will result in Collection(1, 2, 3) aka `Collection<int, int>`, but the type hints think the resulting types are the same 
// as were passed in the original collection
```

File: src/Contract/Operation/Flattenable.php [hunk 5]
```diff
@@ -19,7 +19,7 @@ interface Flattenable
      *
      * @see https://loophp-collection.readthedocs.io/en/stable/pages/api.html#flatten
      *
-     * @return Collection<mixed, mixed>
+     * @return Collection<TKey, T>
```
Comment (AlexandruGG) [inline src/Contract/Operation/Flattenable.php:22]: same issue here as with the `unwrap`

File: src/Operation/Wrap.php [hunk 6]
```diff
@@ -16,25 +16,20 @@
...
-            (new Normalize())()
-        );
-
-        // Point free style.
-        return $pipe;
+        return
+        /**
+         * @param iterable<TKey, T> $iterable
+         *
+         * @return Generator<int, non-empty-array<TKey, T>>
```
Comment (AlexandruGG) [inline src/Operation/Wrap.php:27]: this will not always be a `non-empty-array` I think? What if the `iterable` is empty?

**Review by AlexandruGG — COMMENTED** [review]
I think we need to be careful not to make things way too restrictive -> I see lots of changes needed in the static analysis checks - these are all potential changes users would need to make in their own codebases.

If some of changes are unavoidable simply due to Psalm 5.4 then we can perhaps use a baseline and introduce them more gradually? Would be easier if we would only do it for a set of operations at a time so we can carefully think about how the change affects their usage.

File: composer.json (hunk 1 above)
Comment (drupol) [reply composer.json:41]: We need the next version, 5.4.1 ;)

**Comment by drupol** [comment]
@AlexandruGG All good for you now?

**Comment by AlexandruGG** [comment]
> @AlexandruGG All good for you now?
> 

I'll give this another review tomorrow 😁

**Comment by AlexandruGG** [comment]
> @AlexandruGG All good for you now?

Hey @drupol, I'm not sure what's changed because still see the main issues I commented about are still present. I'm talking about stuff like Unwrap, Flatten, and generally making things too restrictive with respect to static analysis.

**Comment by drupol** [comment]
Right, I removed the changes from `flatten` and `unwrap`.

**Comment by drupol** [comment]
@AlexandruGG Good for you?

**Review by AlexandruGG — APPROVED** [review]
Thanks, I know this took lots of work!

**Comment by drupol** [comment]
Thanks! Indeed it was painful !

**Pull request merged**